handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1707,50 +1706,43 @@ private void startHandshakeProcessing() {
             executor.execute(new Runnable() {
                 @Override
                 public void run() {
-                    handshake(promise, false);
+                    renegotiateOnEventLoop(promise);
                 }
             });
             return promise;
         }
 
-        handshake(promise, false);
+        renegotiateOnEventLoop(promise);
         return promise;
     }
 
+    private void renegotiateOnEventLoop(final Promise<Channel> newHandshakePromise) {
+        final Promise<Channel> oldHandshakePromise = handshakePromise;
+        if (!oldHandshakePromise.isDone()) {
+            // There's no need to handshake because handshake is in progress already.
+            // Merge the new promise into the old one.
+            oldHandshakePromise.addListener(new FutureListener<Channel>() {
```

Comment:
I know this existed before but maybe just use `PromiseNotifier` ? Less code == better code :)